src/engine/Union.cpp

```diff
@@ -0,0 +1,322 @@
+// Copyright 2018, University of Freiburg,
+// Chair of Algorithms and Data Structures.
+// Author: Florian Kramer (florian.kramer@mail.uni-freiburg.de)
+#include "Union.h"
+
+const size_t Union::NO_COLUMN = std::numeric_limits<size_t>::max();
+
+Union::Union(QueryExecutionContext* qec, std::shared_ptr<QueryExecutionTree> t1,
+             std::shared_ptr<QueryExecutionTree> t2)
+    : Operation(qec) {
+  _subtrees[0] = t1;
+  _subtrees[1] = t2;
+
+  // compute the column origins
+  std::unordered_map<string, size_t> variableColumns = getVariableColumns();
+  _columnOrigins.resize(variableColumns.size(), {NO_COLUMN, NO_COLUMN});
+  for (auto it : variableColumns) {
+    // look for the corresponding column in t1
+    auto it1 = t1->getVariableColumnMap().find(it.first);
+    if (it1 != t1->getVariableColumnMap().end()) {
+      _columnOrigins[it.second][0] = it1->second;
+    } else {
+      _columnOrigins[it.second][0] = NO_COLUMN;
+    }
+
+    // look for the corresponding column in t2
+    auto it2 = t2->getVariableColumnMap().find(it.first);
+    if (it2 != t2->getVariableColumnMap().end()) {
+      _columnOrigins[it.second][1] = it2->second;
+    } else {
+      _columnOrigins[it.second][1] = NO_COLUMN;
+    }
+  }
+}
+
+string Union::asString(size_t indent) const {
+  std::ostringstream os;
+  os << _subtrees[0]->asString(indent) << "\n";
+  for (size_t i = 0; i < indent; ++i) {
+    os << " ";
+  }
+  os << "UNION\n";
+  os << _subtrees[1]->asString(indent) << "\n";
+  return os.str();
+}
+
+size_t Union::getResultWidth() const {
+  // the width depends on how many columns use the same variable
```

Comment:
So you mean the width is the number of unique columns? Maybe reformulate that comment. Also what's a `columnOrigin`?

Reply:
I changed the comment. The columnOrigin vector stores from what column in each of the subtrees the values for a column in the result come.